shared/AppInsightsCore/src/JavaScriptSDK/InstrumentHooks.ts

```diff
@@ -0,0 +1,235 @@
+
+        let funcArgs:IInstrumentCallDetails = {
+            name: aiHook.n,
+            inst: funcThis,
+            ctx: null
+        };
+    
+        let hookCtx = [];
+        let cbArgs:any[] = [funcArgs];
+        _arrLoop((orgArgs as any), (arg) => {
+            cbArgs.push(arg);
+        });
+
+        // Call the pre-request hooks
+        _doCallbacks(hooks, funcArgs, cbArgs, hookCtx, CallbackType.Request);
+
+        // Call the original function was called
+        let theFunc = aiHook.f;
+        try {
+            funcArgs.rslt = theFunc.apply(funcThis, orgArgs);
+        } catch (err) {
+            // Report the request callback
+            funcArgs.err = err;
+            _doCallbacks(hooks, funcArgs, cbArgs, hookCtx, CallbackType.FunctionError);
+
+            // rethrow the original exception so anyone listening for it can catch the exception
+            throw err;
+        }
+
+        // Call the post-request hooks
+        _doCallbacks(hooks, funcArgs, cbArgs, hookCtx, CallbackType.Response);
+
+        return funcArgs.rslt;
+    };
+}
+
+/** @ignore */
+function _getOwner(target:any, name:string, checkPrototype:boolean) {
+    let owner = null;
+    if (target) {
+        if (target.hasOwnProperty(name)) {
+            owner = target;
+        } else if (checkPrototype) {
+            owner = _getOwner(target[strPrototype], name, false);
+        }
+    }
+
+    return owner;
+}
+
+/**
+ * Intercept the named prototype functions for the target class / object
+ * @param target - The target object
+ * @param funcName - The function name
+ * @param callbacks - The callbacks to configure and call whenever the function is called
+ */
+export function InstrumentProto(target:any, funcName:string, callbacks: IInstrumentHooksCallbacks): IInstrumentHook {
+    if (target) {
+        return InstrumentFunc(target[strPrototype], funcName, callbacks, false);
+    }
+
+    return null;
+}
+
+/**
+ * Intercept the named prototype functions for the target class / object
+ * @param target - The target object
+ * @param funcNames - The function names to intercept and call
+ * @param callbacks - The callbacks to configure and call whenever the function is called
+ */
+export function InstrumentProtos(target:any, funcNames:string[], callbacks: IInstrumentHooksCallbacks): IInstrumentHook[] {
+    if (target) {
+        return InstrumentFuncs(target[strPrototype], funcNames, callbacks, false);
+    }
+
+    return null;
+}
+
+/**
+ * Intercept the named prototype functions for the target class / object
+ * @param target - The target object
+ * @param funcName - The function name
+ * @param callbacks - The callbacks to configure and call whenever the function is called
+ * @param checkPrototype - If the function doesn't exist on the target should it attempt to hook the prototype function
+ */
+export function InstrumentFunc(target:any, funcName:string, callbacks: IInstrumentHooksCallbacks, checkPrototype:boolean = true): IInstrumentHook {
```

Comment:
Do these functions make sense in core? Would it not make sense to include them in dynamicproto package?

Reply:
They belong here as DynamicProto doesn't do anything like what these functions are doing.